plugins/outputs/websocket/websocket.go

```diff
@@ -0,0 +1,216 @@
+	if err != nil {
+		return fmt.Errorf("error creating TLS config: %v", err)
+	}
+
+	dialer := &ws.Dialer{
+		Proxy:            http.ProxyFromEnvironment,
+		HandshakeTimeout: time.Duration(w.ConnectTimeout),
+		TLSClientConfig:  tlsCfg,
+	}
+
+	headers := http.Header{}
+	for k, v := range w.Headers {
+		headers.Set(k, v)
+	}
+
+	conn, resp, err := dialer.Dial(w.URL, headers)
+	if err != nil {
+		return fmt.Errorf("error dial: %v", err)
+	}
+	defer func() { _ = resp.Body.Close() }()
+	if resp.StatusCode != http.StatusSwitchingProtocols {
+		return fmt.Errorf("wrong status code while connecting to server: %d", resp.StatusCode)
+	}
+
+	w.conn = conn
+	go w.read(conn)
+
+	return nil
+}
+
+func (w *WebSocket) read(conn *ws.Conn) {
+	defer func() { _ = conn.Close() }()
+	if w.ReadTimeout > 0 {
+		if err := conn.SetReadDeadline(time.Now().Add(time.Duration(w.ReadTimeout))); err != nil {
+			return
+		}
+		conn.SetPingHandler(func(string) error {
+			err := conn.SetReadDeadline(time.Now().Add(time.Duration(w.ReadTimeout)))
+			if err != nil {
+				return err
+			}
+			return conn.WriteControl(ws.PongMessage, nil, time.Now().Add(time.Duration(w.WriteTimeout)))
+		})
+	}
+	for {
+		// Need to read a connection (to properly process pings from a server).
+		_, _, err := conn.ReadMessage()
+		if err != nil {
+			if ws.IsUnexpectedCloseError(err, ws.CloseGoingAway, ws.CloseAbnormalClosure) {
+				w.Logger.Errorf("error reading websocket connection: %v", err)
+			}
+			return
+		}
+		if w.ReadTimeout > 0 {
+			if err := conn.SetReadDeadline(time.Now().Add(time.Duration(w.ReadTimeout))); err != nil {
+				return
+			}
+		}
+	}
+}
+
+// Write writes the given metrics to the destination. Not thread-safe.
+func (w *WebSocket) Write(metrics []telegraf.Metric) error {
+	if w.conn == nil {
+		// Previous write failed with error and ws conn was closed.
+		if err := w.Connect(); err != nil {
+			return err
+		}
+	}
+
+	messageData, err := w.serializer.SerializeBatch(metrics)
+	if err != nil {
+		return err
+	}
+
+	messageType := ws.BinaryMessage
+	if w.UseTextFrames {
+		messageType = ws.TextMessage
+	}
+
+	if w.WriteTimeout > 0 {
+		if err := w.conn.SetWriteDeadline(time.Now().Add(time.Duration(w.WriteTimeout))); err != nil {
+			return fmt.Errorf("error setting write deadline: %v", err)
+		}
+	}
+	err = w.conn.WriteMessage(messageType, messageData)
+	if err != nil {
+		_ = w.conn.Close()
+		w.conn = nil
+		return fmt.Errorf("error writing to connection: %v", err)
+	}
+	return nil
+}
+
+// Close closes the connection. Noop if already closed.
+func (w *WebSocket) Close() error {
+	if w.conn == nil {
+		return nil
+	}
+	err := w.conn.Close()
+	w.conn = nil
+	return err
+}
+
+func newWebSocket() *WebSocket {
+	return &WebSocket{
+		ConnectTimeout: config.Duration(defaultConnectTimeout),
+		WriteTimeout:   config.Duration(defaultWriteTimeout),
+		ReadTimeout:    config.Duration(defaultReadTimeout),
+		Logger:         testutil.Logger{},
```

Comment:
Please do not set the logging-facility manually in the plugin. When you rename `Logger` to `Log` as requested above, this field will be automatically set by telegraf. You only need to set the logging-facility manually (as you did here) for the test-suite.

Reply:
I just blindly changed, thanks for catching

Reply:
Thanks for the quick fixes!